### ISSUE TITLE
Inefficiency in SNI policy parsing

### ISSUE THREAD

**Issue by zwoop** [apache/trafficserver#7607]
It seems that rather than parsing these types of "text" configs once, we parse it over and over again (strcmp()) on each transaction hitting this code:

```
std::string_view
HttpSM::get_outbound_sni() const
{
  using namespace ts::literals;
  ts::TextView zret;
  ts::TextView policy{t_state.txn_conf->ssl_client_sni_policy, ts::TextView::npos};

  if (policy.empty() || !strcmp(policy, "host"_tv)) {
    // By default the host header field value is used for the SNI.
    int len;
    char const *ptr = t_state.hdr_info.server_request.host_get(&len);
    zret.assign(ptr, len);
  } else if (ua_txn && !strcmp(policy, "server_name"_tv)) {
    zret.assign(ua_txn->get_netvc()->get_server_name(), ts::TextView::npos);
  } else if (policy.front() == '@') { // guaranteed non-empty from previous clause
    zret = policy.remove_prefix(1);
  } else {
    // If other is specified, like "remap" and "verify_with_name_source", the remapped origin name is used for the SNI value
    zret.assign(t_state.server_info.name, ts::TextView::npos);
  }
  return zret;
}
````

**Comment by masaori335** [comment]
`set_tls_options` has the same issue (`proxy.config.ssl.client.verify.server.policy`/`proxy.config.ssl.client.verify.server.properties`), I think.
https://github.com/apache/trafficserver/blob/d4fc16f64d3104d3682d496e47f2a53511fa77e0/proxy/http/HttpSM.cc#L4810-L4849

**Comment by SolidWallOfCode** [comment]
The tricky bit here is the value is not strictly an enumeration. It can have an arbitrary value as long as that value starts with "@". In such a case, the "@" is removed and the literal value is used as the outbound SNI. This means a string is required in some cases. What would be needed is an enumeration and a string, with the string set only when it's explicitly specified by a plugin.

**Comment by ywkaras** [comment]
Yup https://github.com/apache/trafficserver/pull/7624/files#diff-c2f4153a7fbd17ff9e386fa7e9460927d1eb82b473bb3ec59e23dcccbdaa2037R456 .

**Comment by ywkaras** [comment]
It's unclear to me how using this function, https://github.com/apache/trafficserver/blob/6b28f19d2b863201f87ed31bc933998f26b55f5b/proxy/http/HttpConfig.cc#L74 , leads to a simpler resolution of this issue.  It is roughly equivalent to just:  https://github.com/ywkaras/trafficserver/blob/b0453b157dba0b4d26bf8856f5f7198e2d9241e6/include/tscpp/util/CodeOrStr.h#L180
